macro-utils/src/generator.rs

```diff
@@ -0,0 +1,337 @@
+    /// A `u8` item
+    U8(u8),
+    /// A `i8` item
+    I8(i8),
+    /// A `u16` item
+    U16(u16),
+    /// A `i16` item
+    I16(i16),
+    /// A `u32` item
+    U32(u32),
+    /// A `i32` item
+    I32(i32),
+    /// A `u64` item
+    U64(u64),
+    /// A `i64` item
+    I64(i64),
+    /// A `str` item
+    Bytes(Vec<u8>),
+}
+
+impl Item {
+    /// Convert an item into raw bytes in the specified `endian`.
+    fn into_bytes(self, endian: Endian) -> Vec<u8> {
+        match endian {
+            Endian::Big => match self {
+                Item::U8(val) => val.to_be_bytes().into(),
+                Item::I8(val) => val.to_be_bytes().into(),
+                Item::U16(val) => val.to_be_bytes().into(),
+                Item::I16(val) => val.to_be_bytes().into(),
+                Item::U32(val) => val.to_be_bytes().into(),
+                Item::I32(val) => val.to_be_bytes().into(),
+                Item::U64(val) => val.to_be_bytes().into(),
+                Item::I64(val) => val.to_be_bytes().into(),
+                Item::Bytes(val) => val,
+            },
+            Endian::Little => match self {
+                Item::U8(val) => val.to_le_bytes().into(),
+                Item::I8(val) => val.to_le_bytes().into(),
+                Item::U16(val) => val.to_le_bytes().into(),
+                Item::I16(val) => val.to_le_bytes().into(),
+                Item::U32(val) => val.to_le_bytes().into(),
+                Item::I32(val) => val.to_le_bytes().into(),
+                Item::U64(val) => val.to_le_bytes().into(),
+                Item::I64(val) => val.to_le_bytes().into(),
+                Item::Bytes(val) => val,
+            },
+        }
+    }
+
+    /// Get the size of this item (in bytes).
+    fn size(&self) -> usize {
+        match *self {
+            Item::U8(_) | Item::I8(_) => mem::size_of::<u8>(),
+            Item::U16(_) | Item::I16(_) => mem::size_of::<u16>(),
+            Item::U32(_) | Item::I32(_) => mem::size_of::<u32>(),
+            Item::U64(_) | Item::I64(_) => mem::size_of::<u64>(),
+            Item::Bytes(ref bytes) => bytes.len(),
+        }
+    }
+}
+
+/// A helper macro to build if-else arms in parse function of `Item`.
+macro_rules! item_parse_arms {
+    (
+        $input:ident,
+        $token_name:ident,
+        ($kw:ident, $token:ident) => $blk:block,
+        $(($kws:ident, $tokens:ident) => $blks:block),*
+        $(,)?
+    ) => {
+        let lookahead = $input.lookahead1();
+
+        if lookahead.peek(kw::$kw) {
+            $input.parse::<kw::$kw>()?;
+            $input.parse::<Token![:]>()?;
+            let $token_name = $input.parse::<$token>()?;
+            $blk
+        }
+        $(
+            else if lookahead.peek(kw::$kws) {
+                $input.parse::<kw::$kws>()?;
+                $input.parse::<Token![:]>()?;
+                let $token_name = $input.parse::<$tokens>()?;
+                $blks
+            }
+        )*
+        else {
+            Err(lookahead.error())
```

Comment:
Can we list all possible keywords in the error message? It will help debug.

Reply:
`lookahead.error()` has done it, this function will record all keywords it peeked, and generate an error to list them.

By the way, you can see `macro-utils/tests/ui/generate/fails/*.rs` for test cases when the macro calling fails, and the `.stderr` files with the same names as `.rs` files include the error messages from compiler.